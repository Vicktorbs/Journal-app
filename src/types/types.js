export const types = {
    login: '[Auth] Login',
    loout: '[Auth] Logout'
}