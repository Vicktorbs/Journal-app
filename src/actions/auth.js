import { firebase, googleAuthProvider } from "../firebase/firebase_config";
import { types } from "../types/types"

export const startLoginWithEmailPassword = (email, password) => {
    return (dispatch) => {
        setTimeout(() => {
            dispatch(login(12345, 'Juan'))
        }, 3500);
    }
}

export const startGoogleLogin = () => {
    return (dispatch) => {
        firebase.auth().signInWithPopup(googleAuthProvider).then(({ user }) => {
            dispatch(login(user.uid, user.displayName))
        })
    }
}

export const login = (uid, displayName) => {
    return {
        type: types.login,
        payload: {
            uid,
            displayName
        }
    }
}