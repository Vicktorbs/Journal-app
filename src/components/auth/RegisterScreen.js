import React from 'react';
import { Link } from 'react-router-dom'
import { useForm } from '../../hooks/useForm';
import validator from 'validator'
import { useDispatch, useSelector } from 'react-redux';
import { removeError, setError } from '../../actions/ui';
import { startRegisterWithEmailPasswordName } from '../../actions/auth';

export const RegisterScreen = () => {
    
    const dispatch = useDispatch();
    const { msgError } = useSelector(state => state.ui);

    const [values, handleInputChange] = useForm({
        name: 'Brenda',
        email: 'yoana@gmail.com',
        password: '123456',
        passwordConfirm: '123456'
    });

    const { name, email, password, passwordConfirm } = values;

    const handleSubmit = (e) => {
        e.preventDefault();
        // console.log(name, email, password, passwordConfirm);
        if (isFormValid()) {
            dispatch(startRegisterWithEmailPasswordName(email, password, name))
        }
    }

    const isFormValid = () => {
        
        if (name.trim().length === 0) {
            dispatch(setError('Name required'));
            // console.log('Name required');
            return false
        } else if (!validator.isEmail(email)) {
            dispatch(setError('Email is not valid'));
            // console.log('Email is not valid');
            return false
        } else if (password !== passwordConfirm || password.length < 5) {
            dispatch(setError('Password shold have at least 5 characters and be eaquals'));
            // console.log('Password shold have at least 5 characters and be eaquals');
            return false
        }

        dispatch(removeError());

        return true
    }

    return (
        <>
            <h2 className="auth_title">Register</h2>

            <form onSubmit={ handleSubmit } className="animate__animated animate__fadeIn animate__faster" >
                {
                    msgError && (
                    <div className="auth__alert-error">
                        {msgError}
                    </div>
                    )
                }
                <input 
                    type="text"
                    placeholder="Name"
                    name="name"
                    className="auth__input"
                    autoComplete="off"
                    value={ name }
                    onChange={ handleInputChange }
                />
                <input 
                    type="text"
                    placeholder="Email"
                    name="email"
                    className="auth__input"
                    autoComplete="off"
                    value={ email }
                    onChange={ handleInputChange }
                />
                <input 
                    type="password"
                    placeholder="Password"
                    name="password"
                    className="auth__input"
                    value={ password }
                    onChange={ handleInputChange }
                />
                <input 
                    type="password"
                    placeholder="Confirm password"
                    name="passwordConfirm"
                    className="auth__input"
                    value={ passwordConfirm }
                    onChange={ handleInputChange }
                />
                <button type="submit" className="btn btn-primary btn-block mb-5">
                    Register
                </button>

                <Link to="/auth/login" className="link">
                    Already registered?
                </Link>
            </form>
        </>
    )
}
