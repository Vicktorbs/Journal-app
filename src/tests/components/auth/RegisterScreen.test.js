import React from 'react';
import { mount } from "enzyme"
import { Provider } from "react-redux"
import { RegisterScreen } from "../../../components/auth/RegisterScreen"
import configureStore from 'redux-mock-store'; 
import thunk from 'redux-thunk';
import { MemoryRouter } from "react-router-dom";
import { types } from '../../../types/types';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initialState = {
    auth: {},
    ui: {
      loading: false,
      msgError: null
    },
    notes: {
      notes: [],
      active: null
    }
}

let store = mockStore(initialState);

const wrapper = mount(
    <Provider store={ store }>
        <MemoryRouter>
            <RegisterScreen />
        </MemoryRouter>
    </Provider>
    )

describe('Test on component <RegisterScreen />', () => {
    
    test('should show itself corectly', () => {
        expect(wrapper).toMatchSnapshot();
    })
    
    test('should dispatch the respective action', () => {
        const emailField = wrapper.find('input[name="email"]');

        emailField.simulate('change', {
            target: {
                value: '',
                name: 'email'
            }
        })

        wrapper.find('form').simulate('submit', {
            preventDefault(){}
        })
        const actions = store.getActions();
        // console.log(actions);

        expect(actions[0]).toEqual({
            type: types.uiSetError,
            payload: 'Email is not valid'
        })
    })
    
    test('should show the alert box with the error message', () => {
        const initialState = {
            auth: {},
            ui: {
              loading: false,
              msgError: 'Email is not valid'
            },
            notes: {
              notes: [],
              active: null
            }
        }
        
        const store = mockStore(initialState);
        
        const wrapper = mount(
            <Provider store={ store }>
                <MemoryRouter>
                    <RegisterScreen />
                </MemoryRouter>
            </Provider>
        )

        expect(wrapper.find('.auth__alert-error').exists()).toBeTruthy();
        expect(wrapper.find('.auth__alert-error').text().trim()).toBe(initialState.ui.msgError);
    })
    
})
