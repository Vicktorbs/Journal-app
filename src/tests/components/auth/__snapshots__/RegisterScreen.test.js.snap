// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on component <RegisterScreen /> should show itself corectly 1`] = `
Array [
  <h2
    className="auth_title"
  >
    Register
  </h2>,
  <form
    className="animate__animated animate__fadeIn animate__faster"
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input"
      name="name"
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value="Brenda"
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      type="text"
      value="yoana@gmail.com"
    />
    <input
      className="auth__input"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value="123456"
    />
    <input
      className="auth__input"
      name="passwordConfirm"
      onChange={[Function]}
      placeholder="Confirm password"
      type="password"
      value="123456"
    />
    <button
      className="btn btn-primary btn-block mb-5"
      type="submit"
    >
      Register
    </button>
    <a
      className="link"
      href="/auth/login"
      onClick={[Function]}
    >
      Already registered?
    </a>
  </form>,
]
`;
